tutorials/arvo/cli-tutorial.md

```diff
@@ -0,0 +1,463 @@
+ * `dbug` is a library of debugging tools. Why do we need this?
+ * `default-agent` contains a Gall agent core with minimal implementations of
+   required Gall arms.
+
+```hoon
+|%
++$  state-0  [%0 ~]
++$  command  ~
+::
++$  card  card:shoe
+--
+```
+The types used by the app.
+
+`$state-0` stores the state of the app, which is null as there is no state to
+keep track of. It is good practice (required?) to have a type for state anyways
+in case the app is made stateful at a later time, and indeed we will do this
+when we modify this app later in the tutorial.
+
+`$command` is typically a set of tagged union types that are possible
+commands that can be entered by the user. Since this app only supports one
+command, it is unnecessary for it to have any associated data, thus the command
+is represented by `~`.
+
+In a non-trivial context, a `$command` is given by `[%name
+data]`, where `%name` is the identifier for the type of command and `data` is
+a type or list of types that contain data needed to execute the command. See
+`app/chat-cli.hoon` for examples of commands, such as `[%say letter:store]` and
+`[%delete path]`.
+
+`$card` is either an ordinary Gall agent `card` or a `%shoe` `card`, which takes
+the form `[%shoe sole-ids=(list @ta) effect=shoe-effect]`. A `%shoe` `card` is
+sent to all `sole`s listed in `sole-ids`, instructing them to implement the
+action specified by `effect`. Here we can
+reference `card:shoe` because of `/+  shoe` at the beginning of the app.
+
+```hoon
+=|  state-0
+=*  state  -
+::
+```
+Pin the bunt value of `state-0` to the head of the subject, then give it the
+macro `state`. The `-` here is a lark expression referring to the head of the
+subject. This allows us to use `state` to refer to the state elsewhere in the
+code no matter what version we're using.
+
+```hoon
+%+  verb  |
+%-  agent:dbug
+^-  agent:gall
+%-  (agent:shoe command)
+^-  (shoe:shoe command)
+```
+This sequence of commands forges a `agent:gall` core that
+possess may arms of the sort defined in `verb,` `agent:dbug`, and `agent:shoe`.
+This sequence of commands tells the compiler that the following core is a Gall
+agent core that may possess arms of the sort defined in `verb`, `agent:dbug`,
+and `agent:shoe`. These cores tell the compiler what sort of form these arms
+must have, and ensure that the arms that are not standard Gall arms are put into
+the context so that the resulting core nests within `agent:gall`.
+
+To be more precise, going from bottom to top, we cast what follows to a
+`(shoe:shoe command)` app core,
+pass that into a gate that transforms it into a `(agent:shoe command)` core, which is then cast
+as an `agent:gall` and passed into the `agent:dbug` gate, which
+endows the resulting core (stil a valid Gall agent core) with additional arms
+useful for debugging purposes. Finally, the `verb` gate (imported with `/+
+verb`) is called, which allows the agent to print what it is doing.
```

Comment:
This is a bit misleading, and incorrect on some counts.

The casts are merely type annotations. The way this is written makes it sounds like some transformation happens when we cast _to_ these types, but they're just reminders that we are writing/producing the types specified.

The _call_ to `+agent:shoe` _produces_ a standard gall agent, and from there on everything we do operates on standard gall agents. There are no extra arms, just agents wrapping around agents and adding their own logic.

Reply:
Oops, missed linking this yesterday: https://github.com/urbit/docs/pull/912/commits/cae94e6de188b62721bb40412c37bf52b0aca6fc#diff-6d8f36f0b87617684df85577b6a7144dL298